cli/src/main/java/org/aya/cli/parse/AyaProducer.java

```diff
@@ -366,11 +366,94 @@ public QualifiedID visitQualifiedId(AyaParser.QualifiedIdContext ctx) {
         sourcePosOf(forall), false,
         visitForallTelescope(forall.tele()).view(),
         visitExpr(forall.expr()));
+      case AyaParser.DoContext doCtx -> visitDo(doCtx);
+      case AyaParser.IdiomContext idmCtx -> {
+        if (idmCtx.idiomBlock().barredExpr() == null)
+          yield new Expr.UnresolvedExpr(sourcePosOf(idmCtx), "empty");
+        yield visitIdiomBlock(idmCtx.idiomBlock());
+      }
+      case AyaParser.ArrayContext arrCtx -> {
+        if (arrCtx.exprList() == null)
+          yield new Expr.UnresolvedExpr(sourcePosOf(arrCtx), "nil");
+        yield visitArr(arrCtx);
+      }
       // TODO: match
       default -> throw new UnsupportedOperationException("TODO: " + ctx.getClass());
     };
   }
 
+  private @NotNull Expr visitArr(AyaParser.ArrayContext ctx) {
+    var consArg = new Expr.NamedArg(true, new Expr.UnresolvedExpr(sourcePosOf(ctx), ":<"));
+    var nilArg = new Expr.NamedArg(true, new Expr.UnresolvedExpr(sourcePosOf(ctx), "nil"));
+    var args = ImmutableSeq.from(ctx.exprList().expr()).view()
+      .map(exprCtx -> new Expr.NamedArg(true, visitExpr(exprCtx)))
+      .flatMap(expr -> ImmutableSeq.of(expr, consArg))
+      .appended(nilArg)
+      .toImmutableSeq();
+    return new Expr.BinOpSeq(sourcePosOf(ctx), args);
+  }
+
+  /**
+   * Warning: the parser cannot enforce left associativity at this stage
+   */
+  private @NotNull Expr visitIdiomBlock(AyaParser.IdiomBlockContext ctx) {
+    var orArg = new Expr.NamedArg(true, new Expr.UnresolvedExpr(sourcePosOf(ctx), "<*>"));
```

Comment:
Does this mean users should declare the `infixl` for `<*>`?

Reply:
But indeed you can enforce left-assoc by putting binary args in a `newBinOPScope`, for example, the code `a <*> b <*> c` can be parsed as `newBinOPScope(newBinOPScope(a, <*>, b), <*>, c)` which will ignore the assiciativity of `<*>` since there's only two operands.

Reply:
I prefer local open btw